precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+use frame_support::weights::Weight;
+use frame_support::{codec::DecodeLimit as _, traits::Get};
+use frame_support::{
+    dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+    traits::ConstU32,
+};
+use frame_system::Config;
+use pallet_evm::GasWeightMapping;
+use pallet_evm_precompile_dispatch::DispatchValidateT;
+use precompile_utils::prelude::{revert, BoundedBytes, RuntimeHelper, UnboundedBytes};
+use precompile_utils::EvmResult;
+use sp_core::{crypto::AccountId32, H160, H256};
+use sp_io::hashing::keccak_256;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const LOG_TARGET: &str = "precompile::dispatch-lockdrop";
+
+// ECDSA PublicKey
+type ECDSAPublic = ConstU32<64>;
+
+// `DecodeLimit` specifies the max depth a call can use when decoding, as unbounded depth
+// can be used to overflow the stack.
+// Default value is 8, which is the same as in XCM call decoding.
+pub struct DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit = ConstU32<8>>(
+    PhantomData<(Runtime, DispatchValidator, DecodeLimit)>,
+);
+
+#[precompile_utils::precompile]
+impl<Runtime, DispatchValidator, DecodeLimit>
+    DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit>
+where
+    Runtime: pallet_evm::Config,
+    <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+    Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    <Runtime as Config>::AccountId: IsType<AccountId32>,
+    <Runtime as Config>::AccountId: From<[u8; 32]>,
+    DispatchValidator:
+        DispatchValidateT<<Runtime as Config>::AccountId, <Runtime as Config>::RuntimeCall>,
+    DecodeLimit: Get<u32>,
+{
+    #[precompile::public("dispatch_lockdrop_call(bytes,bytes)")]
+    fn dispatch_lockdrop_call(
+        handle: &mut impl PrecompileHandle,
+        call: UnboundedBytes,
+        pubkey: BoundedBytes<ECDSAPublic>,
+    ) -> EvmResult<bool> {
+        log::trace!(
+            target: LOG_TARGET,
+            "raw arguments: call: {:?}, pubkey: {:?}",
+            call,
+            pubkey
+        );
+
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+
+        // Record the cost of the call to ensure there is no free execution
+        handle.record_cost(Runtime::GasWeightMapping::weight_to_gas(
+            Weight::from_parts(1_000_000u64, 0),
+        ))?;
+
+        // Ensure that the caller matches the public key
+        if caller != Self::get_evm_address_from_pubkey(pubkey.as_bytes()) {
+            let message: &str = "caller does not match the public key";
+            log::trace!(target: LOG_TARGET, "{}", message);
+            return Err(revert(message));
+        }
+
+        // Derive the account id from the public key
+        let origin = Self::get_account_id_from_pubkey(pubkey.as_bytes())
+            .ok_or(revert("could not derive AccountId from pubkey"))?;
```

Comment:
Is it possible to cover this case in unit tests?

Reply:
Added a tests for it. Using the same pubkey & AccountId as from old pallet eth-sig https://github.com/AstarNetwork/Astar/blob/9395ce0c50da8fc92b368be37403d839e5a17448/pallets/custom-signatures/src/tests.rs#L267